### ISSUE TITLE
execution stucks in chrome headless

### ISSUE THREAD

**Issue by vasilyevi** [DevExpress/testcafe#7898]
### What is your Scenario?

Trying to execute tests with concurrency 3 in chrome headless mode

### What is the Current behavior?

first 3 tests marked as failed and tight after the execution stucks

### What is the Expected behavior?

next tests have to be executed

### What is your public website URL? (or attach your complete example)

https://staging.guidetoiceland.is

### What is your TestCafe test code?

```
import { randomInt } from 'fp-ts/lib/Random';
import { ClientFunction, RequestHook, RequestMock, Selector } from 'testcafe';
import moment from 'moment';

fixture`A set of examples that illustrate how to use TestCafe API`

export async function removeTargetAttributes() {
  return ClientFunction(() => {
    const selector = document.querySelectorAll("div[data-testid='availability-cta'] a");
    selector.forEach(el => el.removeAttribute("target"));
  })();
}

async function waitForVisible(t, selector, timeout = 10000) {
  await t
    .expect(selector.with({ timeout: 10000 }).visible)
    .ok(`element '${selector[Object.getOwnPropertySymbols(selector)[0]].fn}' is not visible`, {
      timeout,
    });
}

async function waitForExist(t, selector, timeout = 10000) {
  await t
    .expect(selector.exists)
    .ok(`element '${selector[Object.getOwnPropertySymbols(selector)[0]].fn}' is not exist`, {
      timeout,
    });
}

async function waitForNotExist(t, selector, timeout = 10000) {
  await t
    .expect(selector.exists)
    .notOk(`element '${selector[Object.getOwnPropertySymbols(selector)[0]].fn}' is exist`, {
      timeout,
    });
}

async function waitForEnabled(t, selector, timeout = 10000) {
  await t
      .expect(selector.hasAttribute("disabled"))
      .notOk(`element '${selector[Object.getOwnPropertySymbols(selector)[0]].fn}' is not enabled`, {
        timeout,
      });
}

async function isVisible(selector, timeout = 10000) {
  const result = await selector.with({ timeout }).visible;
  return result;
}

async function beforeHook(t) {
  await t.navigateTo("https://staging.guidetoiceland.is")
  if (t.browser.alias.includes("headless")) {
    await t.resizeWindow(1900, 1000);
  } else {
    await t.maximizeWindow();
  }
  await waitForVisible(t, Selector("[data-id='components_ui_Tabs_RoundedTabs_TabsWrapper']"));
  await waitForExist(t, Selector(":is(#coverCarouselfrontCover,[data-id='components_ui_ImageCarousel_ImageGalleryCarousel_Wrapper'])"));
  await waitForVisible(t, Selector("#contactUsButtonWrapper button"));
  await waitForNotExist(t, Selector("#navBarCurrencyPicker").withText("• • •"));
}

export const productTypes = {
  Trips: "Trips",
};

export function randomInteger(min, max) {
  const rnd = randomInt(min, max)();
  return rnd;
}

async function activateSearch(t, searchType) {
  switch (searchType) {
    case productTypes.Trips:
      await t.click(Selector("[data-testid='trips-tab-label']"));
      break;
    default:
      throw Error("Undefined type of Search");
  }
}

async function search(t, searchType) {
  if (!(await isVisible(Selector("#nextMonth"), 3000))) {
    await t.click(Selector(`:is([data-testid='datePicker'],[data-testid='datePickerstarting-location-id']) svg:first-of-type`));
  }
  await t.click(Selector(".DayPicker-Month .DayPicker-Week div[class*='DayPicker-Day'][aria-disabled='false']").nth(20));
  await t.click(Selector(".DayPicker-Month .DayPicker-Week div[class*='DayPicker-Day'][aria-disabled='false']").nth(25));
  switch (searchType) {
    case productTypes.Trips:
      await t.typeText(Selector(":is([data-testid='trips-location-idAutocomplete'],[data-testid='starting-location-idAutocomplete']"), "Reykjavik")
      await t.click(Selector("div[data-testid$='AutocompleteWrapper'] a").withText(/Reykjavik/));
      await t.click(Selector('[data-testid="searchButton"]'));
      await waitForVisible(t, Selector("[data-id='components_features_SearchPage_TourSearchAndCategoryContainer_Row']"));
      await waitForVisible(t, Selector("[data-testid='imageWrapper']"));
      await waitForNotExist(t, Selector("[data-id='components_ui_Search_LoadingBar_Wrapper']"));
      await t.eval(() => {
        const selector = document.querySelectorAll("a[target]");
        selector.forEach(el => el.removeAttribute("target"));
      });
      await t.takeScreenshot();
      const rnd = randomInteger(1, (await Selector("[data-testid='imageWrapper']").count) - 1);
      await removeTargetAttributes();
      await t.click(Selector("[data-testid='imageWrapper']").nth(rnd));
      await waitForVisible(t, Selector("#pageContentContainer"));
      await waitForVisible(t, Selector("[data-testid='continueBookButton'],#stickyFooter button")
        .withText(/(Continue to book|Continue|Book Now|Add to Cart)/)
        .filterVisible(),
        30000);
      await t.click(Selector("[data-testid='continueBookButton'],#stickyFooter button")
        .withText(/(Continue to book|Continue|Book Now|Add to Cart)/)
        .filterVisible());
      await waitForEnabled(t, Selector("#paymentFormButton"), 30000);
      break;
    default:
      throw Error("Undefined type of Search");
  }
}

async function searchForProduct(t, product) {
  await activateSearch(t, product);
  await search(t, product);
}

test('GTI trip', async t => {
  console.log(`${moment().format("HH:MM:ss")} GTI trip is stated`);
  await beforeHook(t);
  await searchForProduct(t, productTypes.Trips);
  console.log(`${moment().format("HH:MM:ss")} GTI trip is finished`);
});

test('GTI trip2', async t => {
  console.log(`${moment().format("HH:MM:ss")} GTI trip2 is stated`);
  await beforeHook(t);
  await searchForProduct(t, productTypes.Trips);
  console.log(`${moment().format("HH:MM:ss")} GTI trip2 is finished`);
});

test('GTI trip3', async t => {
  console.log(`${moment().format("HH:MM:ss")} GTI trip3 is stated`);
  await beforeHook(t);
  await searchForProduct(t, productTypes.Trips);
  console.log(`${moment().format("HH:MM:ss")} GTI trip3 is finished`);
});

test('GTI trip4', async t => {
  console.log(`${moment().format("HH:MM:ss")} GTI trip4 is stated`);
  await beforeHook(t);
  await searchForProduct(t, productTypes.Trips);
  console.log(`${moment().format("HH:MM:ss")} GTI trip4 is finished`);
});

test('GTI trip5', async t => {
  console.log(`${moment().format("HH:MM:ss")} GTI trip5 is stated`);
  await beforeHook(t);
  await searchForProduct(t, productTypes.Trips);
  console.log(`${moment().format("HH:MM:ss")} GTI trip5 is finished`);
});

test('GTI trip6', async t => {
  console.log(`${moment().format("HH:MM:ss")} GTI trip6 is stated`);
  await beforeHook(t);
  await searchForProduct(t, productTypes.Trips);
  console.log(`${moment().format("HH:MM:ss")} GTI trip6 is finished`);
});
```

### Your complete configuration file

```
const requestLogger = require("./tests/requestLogger");

module.exports = {
    screenshots: {
        path: "./artifacts/screenshots",
        takeOnFails: true,
        pathPattern: "${DATE}_${TIME}/${TEST}/${FILE_INDEX}_${QUARANTINE_ATTEMPT}.png",
        thumbnails: false
    },
    disableMultipleWindows: true,
    browsers: "chrome:headless",
    concurrency: 3,
    skipJsErrors: true,
    pageLoadTimeout: 40000,
    pageRequestTimeout: 45000,
    selectorTimeout: 15000,
    disableNativeAutomation: false,
    hooks: {
        request: requestLogger,
    },
}
```

### Steps to Reproduce

1. Run tests

### TestCafe version

3.1.0-rc.1

### Node.js version

v16.19.0

### Command-line arguments

npx testcafe

### Platform(s) and version(s)

MacOS 13.4.1

**Comment by vasilyevi** [comment]
I do have one more weird case on our prod env, it requires `addRequestHooks` in the beginning of the tests. The behavior is the following. Tests start, stuck and re-run automatically without any errors.
Please reach out to me directly as I cannot post the info here. Telegram is @f_the_w

**Comment by Aleksey28** [comment]
Hi @vasilyevi,

I managed to reproduce your example in TestCafe v3.0.0 in native automation mode independent of headless mode. It isn't stuck - it tries to find the `Selector(".DayPicker-Month .DayPicker-Week div[class*='DayPicker-Day'][aria-disabled='false']")` selector, but there is no selector at that moment. Also, you have a huge `selectorTimeout` timeout. That is why it waits so long before throwing an error. The initial problem is that TestCafe doesn't open the calendar in the previous step.

Please create a simple sample with the implementation of opening this calendar and a test that reproduces this error in that sample.

> I do have one more weird case on our prod env, it requires addRequestHooks in the beginning of the tests.

You can contact our support team ([support@devexpress.com](mailto:support@devexpress.com)) and share your example. Please make this sample as simple as possible. Also, note that if your website is private and requires authentication, we can't access it. Our policy prevents us from accessing customers' internal resources without written approval from the person who owns the server/web resource. If you want us to research the issue further, we need access to the server/web resource. Please ask the website owner to send us a written confirmation. It must permit DevExpress personnel to remotely access the website and its internal resources for research, testing, and debugging purposes.

**Comment by vasilyevi** [comment]
@Aleksey28, not sure you reproduced it. My execution passes that selector and after just doing nothing for 10+ minutes

**Comment by vasilyevi** [comment]
And works fine without headless

**Comment by vasilyevi** [comment]
Here is a screenshot
I ran the tests, they successfully passed the mentioned selector, but stucked without any issue and were automatically executed again. Without any errors too, the quarantineMode is off. I don't understand what happens. Again, it works fine without headless
![image](https://github.com/DevExpress/testcafe/assets/28949557/32e2e178-f541-48cf-aeef-b8177f867a44)

**Comment by Aleksey28** [comment]
I tried to run your example on:

* MacOS Monterey 12
* Chrome 114
* Node v16.20.1
* testcafe v3.1.0-rc.3


Tests finished in 2m 29s. Two of them failed on row 41 with the error that testcafe cannot find `Selector("#paymentFormButton")`. Please review your example again. Also, pay attention that I used a simple `RequestLogger` since you didn't share yours.

**Comment by vasilyevi** [comment]
@Aleksey28 
The only one difference I see is the macOS version, I have 13.4.1
 I also just updated the node to v18.17.0. and now the first 3 tests are failing as expected but 4, 5 and 6 tests are not running at all. It just stucks right after first 3 are failed.
 
 The `RequestLogger ` is following
 ```
const { RequestLogger } = require("testcafe");

module.exports = new RequestLogger(/client-api/, {
  logResponseBody: true,
  logRequestHeaders: true,
  stringifyResponseBody: true,
});
```

Now I'm also using rc.3

![image](https://github.com/DevExpress/testcafe/assets/28949557/15c72b6c-bb09-4c61-b470-a37ccd9d12e4)

**Comment by vasilyevi** [comment]
Is there any ways to add some logs or something at least to understand what exactly happens?

**Comment by vasilyevi** [comment]
@Aleksey28  I just updated to 3.1.0. The issue is still present.
It stucks right after clicking by link with target attribute. Which actually have to be removed in previous steps. And these attributes were actually removed. After few mins testcafe reruns the same test again by some reason. Anfter few times it fails the whole execution
Check it out. 

![image](https://github.com/DevExpress/testcafe/assets/28949557/8053ed12-1034-4372-a253-792b7799cbe7)

here is the changed a bit code of `search` function 
```
async function search(t, searchType) {
  if (!(await isVisible(Selector("#nextMonth"), 3000))) {
    await t.click(Selector(`:is([data-testid='datePicker'],[data-testid='datePickerstarting-location-id']) svg:first-of-type`));
  }
  await t.click(Selector(".DayPicker-Month .DayPicker-Week div[class*='DayPicker-Day'][aria-disabled='false']").nth(20));
  await t.click(Selector(".DayPicker-Month .DayPicker-Week div[class*='DayPicker-Day'][aria-disabled='false']").nth(25));
  switch (searchType) {
    case productTypes.Trips:
      await t.typeText(Selector(":is([data-testid='trips-location-idAutocomplete'],[data-testid='starting-location-idAutocomplete']"), "Reykjavik")
      await t.click(Selector("div[data-testid$='AutocompleteWrapper'] a").withText(/Reykjavik/));
      await t.click(Selector('[data-testid="searchButton"]'));
      await waitForVisible(t, Selector("[data-id='components_features_SearchPage_TourSearchAndCategoryContainer_Row']"));
      await waitForVisible(t, Selector("[data-testid='imageWrapper']"), 30000);
      await waitForNotExist(t, Selector("[data-id='components_ui_Search_LoadingBar_Wrapper']"));
      await t.eval(() => {
        const selector = document.querySelectorAll("a[target]");
        selector.forEach(el => el.removeAttribute("target"));
      });
      await t.takeScreenshot();
      const rnd = randomInteger(1, (await Selector("[data-testid='imageWrapper']").count) - 1);
      console.log("the time now is: " + moment().format("HH:MM:ss"));
      console.log(moment().format("HH:MM:ss") + " next step is removeTargetAttributes to remove target attribute");
      await removeTargetAttributes();
      await t.takeScreenshot();
      //console.log(`DOM is: '${await getDom()}'`);
      console.log(moment().format("HH:MM:ss") + " next step is click on link");
      await t.click(Selector("[data-testid='imageWrapper']").nth(rnd));
      console.log(moment().format("HH:MM:ss") + " it clicked");
      await waitForVisible(t, Selector("#pageContentContainer"));
      await waitForVisible(t, Selector("[data-testid='continueBookButton'],#stickyFooter button")
        .withText(/(Continue to book|Continue|Book Now|Add to Cart)/)
        .filterVisible(),
        30000);
      await t.click(Selector("[data-testid='continueBookButton'],#stickyFooter button")
        .withText(/(Continue to book|Continue|Book Now|Add to Cart)/)
        .filterVisible());
      await waitForEnabled(t, Selector("#paymentFormButton"), 30000);
      break;
    default:
      throw Error("Undefined type of Search");
  }
}
```

**Comment by vasilyevi** [comment]
here is without headless....
![image](https://github.com/DevExpress/testcafe/assets/28949557/586897c9-6430-4b50-bd9e-8c666e8c5180)

**Comment by vasilyevi** [comment]
same on ubuntu

**Comment by vasilyevi** [comment]
I'll send you details to support@devexpress.com please, check it out

**Comment by AlekseyKravtsov17** [comment]
Hi @Aleksey28,

I have the similar case with identical testcafe behavior. And it's a huge blocker for me.

**Comment by cesnietor** [comment]
Hi, we also started seeing issues with Chrome just recently. changing  to Firefox (firefox --headless) passess them just fine but we'd like to test different browsers specially Chrome.

**Comment by Aleksey28** [comment]
We appreciate you taking the time to share the information about this issue. We replicated it, and it is currently in our internal queue. Please note that the research may take time. We'll update this thread once we have news.

**Comment by miherlosev** [comment]
Hi

I see that you are using the `disableMultipleWindows: true` option here. There is an [issue](https://github.com/DevExpress/testcafe/issues/7916) with it. Let's wait until the issue is resolved and then come back to this.

**Comment by vasilyevi** [comment]
@miherlosev I'm able to reproduce it even with `disableMultipleWindows: false` it DOES NOT somehow related to the current issue. Please, prioritize it as hight priority as it's a full blocker for me and seems like for a lot of people.

**Comment by AlexKamaev** [comment]
I confirm that the issue is not related to the `disableMultipleWindows` options.
I'm researching the issue. However, I do not have any results yet.

**Comment by vasilyevi** [comment]
@AlexKamaev friendly ping :) Do you have any expectations on the issue?

**Comment by AlexKamaev** [comment]
@vasilyevi
I'm still working on the fix. I have some results to show. However, the PR is not ready. If you want, you can check your usage scenario using TestCafe from the following branch:
PR - [https://github.com/DevExpress/testcafe/pull/7939](https://github.com/DevExpress/testcafe/pull/7939)
Branch - [https://github.com/AlexKamaev/testcafe/tree/gh-7898](https://github.com/AlexKamaev/testcafe/tree/gh-7898)

Please note that, in this case, you need to clone the TestCafe repo from the above branch and build it ([https://testcafe.io/documentation/402834/guides/basic-guides/install-testcafe#install-testcafe-from-source](https://testcafe.io/documentation/402834/guides/basic-guides/install-testcafe#install-testcafe-from-source)).


If you check your usage scenario using this TestCafe version, please share your results with us.

**Comment by vasilyevi** [comment]
3.2.0 does not include the fix, does it?

**Comment by AlexKamaev** [comment]
@vasilyevi Version 3.2.0 does not contain the fix. By the way, did you have a chance to check if your usage scenario works with the master branch?

**Comment by vasilyevi** [comment]
@AlexKamaev I haven't checked yet. Which version will include that?

**Comment by AlexKamaev** [comment]
v3.3.0 **will** include the fix. However, you can check if your usage scenario works by using the `master` branch. Here is a guide on how to build and run TestCafe from sources:
[https://testcafe.io/documentation/402834/guides/basic-guides/install-testcafe#install-testcafe-from-source](https://testcafe.io/documentation/402834/guides/basic-guides/install-testcafe#install-testcafe-from-source)

**Comment by vasilyevi** [comment]
@AlexKamaev  Fixed. When do you plan to release 3.3.0?

**Comment by AlexKamaev** [comment]
@vasilyevi
We are working on this. We are planning to release the 3.3.0 version next week if everything is ok.

**Comment by krzysiek-b** [comment]
Have you managed to confirm this issue is really fixed in 3.3.0? I'm using testcafe 3.3.0 and it still hangs. The execution looks always similar. First two tests are executed, third hangs when running headless on first await. When run without headless everything works fine. For now I downgraded to 2.6.2 as this was the last version that worked (works now) for me.

**Comment by AlexKamaev** [comment]
@krzysiek-b 
The issue from the attached example has been fixed. If you have a similar issue, please create a sample that demonstrates the incorrect behavior using the following link: [https://github.com/DevExpress/testcafe/issues/new?assignees=&labels=TYPE%3A+bug&projects=&template=bug_report.yaml](https://github.com/DevExpress/testcafe/issues/new?assignees=&labels=TYPE%3A+bug&projects=&template=bug_report.yaml)
